### PR TITLE
[typescript-resolvers] Add directiveContextExtender option

### PR DESCRIPTION
**Pull request by ludusrusso**
## Description

This PR add the `directiveContextExtender` 

Related #7505 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

I've locally tested this changes with `yarn test`
I've add a new test to check if my contributions is working properly.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

**Test Environment**:
- OS: MacOS 12.1 (Apple Silicon)
- `@graphql-codegen/...`: 
- NodeJS: v16.13.2

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts [hunk 1]
```diff
@@ -932,6 +977,18 @@ export class BaseResolversVisitor<
       const realType = baseType.name.value;
       const parentType = this.schema.getType(parentName);
 
+      let contextName = this._fieldContextTypeMap[`${parentName}.${node.name}`]
+        ? this._fieldContextTypeMap[`${parentName}.${node.name}`].type
+        : 'ContextType';
+
+      for (const directive of node.directives) {
+        const name = directive.name as unknown as string;
```
Comment (ludusrusso) [inline packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:985]: I need some help here, seemps that the type for `node.directives` is not correct!
Can I fix it?

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts [hunk 2]
```diff
@@ -152,6 +153,23 @@ export interface RawResolversConfig extends RawConfig {
    *     rootValueType: ./my-types#MyRootValue
    * ```
    */
+  directiveContextExtender?: Array<string>;
+  /**
+   * @description Use this to set a custom type for a specific field `context` decoreted by a directive.
```
Comment (charlypoly) [inline packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:158]:
```suggestion
   * @description Use this to set a custom type for a specific field `context` decorated by a directive.
```

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts [hunk 3]
```diff
@@ -152,6 +153,23 @@ export interface RawResolversConfig extends RawConfig {
...
+   * You can either use `Field.Path#ContextTypeName` or `Field.Path#ExternalFileName#ContextTypeName`
+   *
+   * @exampleMarkdown
+   * ## Directive Context Extender
+   *
+   * ```yml
+   * plugins
+   *   config:
+   *     directiveContextExtender:
+   *       - myDirecorty#CustomContextExtender
```
Comment (charlypoly) [inline packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:169]:
```suggestion
   *       - myCustomDirectiveName#./my-file#CustomContextExtender
```

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts [hunk 4]
```diff
@@ -152,6 +153,23 @@ export interface RawResolversConfig extends RawConfig {
    *     rootValueType: ./my-types#MyRootValue
    * ```
    */
+  directiveContextExtender?: Array<string>;
```
Comment (charlypoly) [inline packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:156]: Why not `directiveContextTypes`?

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts [hunk 5]
```diff
@@ -1021,6 +1054,21 @@ export class BaseResolversVisitor<
     };
   }
 
+  private getContextType(parentName: string, node: FieldDefinitionNode): string {
+    let contextType = this._fieldContextTypeMap[`${parentName}.${node.name}`]
+      ? this._fieldContextTypeMap[`${parentName}.${node.name}`].type
+      : 'ContextType';
+
+    for (const directive of node.directives) {
```
Comment (charlypoly) [inline packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:1062]: If a given node matches both a `fieldContextTypes` and multiple `directiveContextExtender`,
Isn't it supposed to merge the context types together?
Also, we need a test to cover this scenario 👀 

**Example**

```gql
# schema.graphql
type Mutation {
  notAuthMutaiton(): [NotAuthRes!]!
  loggedMutation (content: String!): loggedRes! @isLoggedIn
}

type Query {
  loggedQuery: LoggedRes!  @isLoggedIn
}
```

```yaml
# codegen.yml
generates:
  ./src/graphql/resolvers/types.generated.ts:
    plugins:
      - typescript
      - typescript-resolvers
    config:
      contextType: ./context.ts#Context
    fieldContextTypes:
         - Query.loggedQuery#./my-file#ExternalContextType
     directiveContextExtender:
      - isLoggedIn: ./context.ts#LoggedContext
```

**Review by charlypoly — CHANGES_REQUESTED** [review]

**Comment by charlypoly** [comment]
Hi @ludusrusso,

Thank you for your contribution!

I've left some comments on naming and doc, and also a specific design question around using `fieldContextTypes` along with `directiveContextExtender `

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts (hunk 4 above)
Comment (ludusrusso) [reply packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:156]: like it!

**Comment by ludusrusso** [comment]
HI @charlypoly! Thanks for help! I've integrated your suggestins and I've added more test cases as suggested.

I'd like also tu add a test case with multiple directives on the same filed, but if I try to uodate the common schema for tests, I breaks too much test. How do you suggest to handle this?

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts (hunk 5 above)
Comment (ludusrusso) [reply packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:1067]: I've add tests to check this!

**Comment by ludusrusso** [comment]
How do I fix the vercel failuer?

**Review by charlypoly — CHANGES_REQUESTED** [review]

File: packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts [hunk 6]
```diff
@@ -152,6 +153,23 @@ export interface RawResolversConfig extends RawConfig {
    *     rootValueType: ./my-types#MyRootValue
    * ```
    */
+  directiveContextTypes?: Array<string>;
+  /**
+   * @description Use this to set a custom type for a specific field `context` decorated by a directive.
```
[the remark this answers is not in the document]
Comment (ludusrusso) [reply packages/plugins/other/visitor-plugin-common/src/base-resolvers-visitor.ts:158]: Done. Do you like this or I should also add an example?

**Comment by charlypoly** [comment]
@ludusrusso Github Actions is currently having an incident 💥 

I'll merge the PR tomorrow (when GH Actions will be working again)

**Pull request merged**